Source/DafnyCore/Resolver/ExpressionTester.cs

```diff
@@ -350,36 +350,55 @@ public class ExpressionTester {
   }
 
   public static bool IsTypeTestCompilable(TypeTestExpr tte) {
-    Contract.Requires(tte != null);
-    var fromType = tte.E.Type;
-    if (fromType.IsSubtypeOf(tte.ToType, false, true)) {
-      // this is a no-op, so it can trivially be compiled
+    return IsTypeTestCompilable(tte.E.Type, tte.ToType);
+  }
+
+  public static bool IsTypeTestCompilable(Type fromType, Type toType) {
+    if (fromType.IsSubtypeOf(toType, false, true)) {
+      // this is a no-op or a simple null comparison, so it can trivially be compiled
       return true;
     }
 
-    // TODO: It would be nice to allow some subset types in test tests in compiled code. But for now, such cases
-    // are allowed only in ghost contexts.
-    var udtTo = (UserDefinedType)tte.ToType.NormalizeExpandKeepConstraints();
-    if (udtTo.ResolvedClass is SubsetTypeDecl and not NonNullTypeDecl) {
-      return false;
+    var udtTo = toType.NormalizePastUnconstrainedTypes() as UserDefinedType;
+    if (udtTo == null) {
+      Contract.Assert(toType.NormalizeToAncestorType() is BoolType or CharType or IntType or BitvectorType or BigOrdinalType or RealType);
+      return true;
+    }
+    if (udtTo.ResolvedClass is SubsetTypeDecl or NewtypeDecl) {
+      var declWithConstraints = (RedirectingTypeDecl)udtTo.ResolvedClass;
+      if (!declWithConstraints.ConstraintIsCompilable) {
+        return false;
+      }
+      // okay, so the constraints are compilable, but we still need to check the injectivity property below
```

Comment:
> the injectivity property

This only gets introduced a few lines later. I think it's better not to have this forward reference. Code should read like a book !

Reply:
I added a description of the method in a `<summary>` comment.